components/brave_ads/core/internal/serving/permission_rules/permission_rules.cc

```diff
@@ -0,0 +1,350 @@
+
+#include <vector>
+
+#include "base/time/time.h"
+#include "brave/components/brave_ads/core/internal/account/issuers/issuers_util.h"
+#include "brave/components/brave_ads/core/internal/account/tokens/confirmation_tokens/confirmation_tokens_util.h"
+#include "brave/components/brave_ads/core/internal/browser/browser_manager.h"
+#include "brave/components/brave_ads/core/internal/catalog/catalog_util.h"
+#include "brave/components/brave_ads/core/internal/client/ads_client_util.h"
+#include "brave/components/brave_ads/core/internal/common/logging_util.h"
+#include "brave/components/brave_ads/core/internal/common/platform/platform_helper.h"
+#include "brave/components/brave_ads/core/internal/common/time/time_constraint_util.h"
+#include "brave/components/brave_ads/core/internal/flags/did_override/did_override_command_line_flag_util.h"
+#include "brave/components/brave_ads/core/internal/flags/environment/environment_flag_util.h"
+#include "brave/components/brave_ads/core/internal/serving/permission_rules/permission_rule_feature.h"
+#include "brave/components/brave_ads/core/internal/settings/settings.h"
+#include "brave/components/brave_ads/core/internal/tabs/tab_manager.h"
+#include "brave/components/brave_ads/core/internal/units/inline_content_ad/inline_content_ad_feature.h"
+#include "brave/components/brave_ads/core/internal/units/new_tab_page_ad/new_tab_page_ad_feature.h"
+#include "brave/components/brave_ads/core/internal/units/promoted_content_ad/promoted_content_ad_feature.h"
+#include "brave/components/brave_ads/core/internal/user/user_attention/user_activity/user_activity_scoring_util.h"
+#include "brave/components/brave_ads/core/internal/user/user_interaction/ad_events/ad_event_cache_util.h"
+#include "brave/components/brave_ads/core/public/account/confirmations/confirmation_type.h"
+#include "brave/components/brave_ads/core/public/units/ad_type.h"
+#include "brave/components/brave_ads/core/public/units/notification_ad/notification_ad_feature.h"
+#include "brave/components/brave_ads/core/public/units/search_result_ad/search_result_ad_feature.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace brave_ads {
+
+namespace {
+
+// A one hour constraint for a given permission rule
+constexpr base::TimeDelta kOneHour = base::Hours(1);
+
+// A one day constraint for a given permission rule
+constexpr base::TimeDelta kOneDay = base::Days(1);
+
+// The minimum wait time cap for minimum wait permission rules
+constexpr int kMinimumWaitTimeCap = 1;
+
+bool IsAdTypeInTimeRangeAllowed(const AdType& type,
+                                base::TimeDelta constraint,
+                                size_t cap) {
+  const std::vector<base::Time> history =
+      GetCachedAdEvents(type, ConfirmationType::kServed);
+
+  return DoesHistoryRespectRollingTimeConstraint(history, constraint,
+                                                 /*cap=*/cap);
+}
+
+}  // namespace
+
+bool ShouldAllowFullScreenMode() {
```

Comment:
Thanks for making these free functions; we will need to consider better naming because `ShouldAllowFullScreenMode` sounds like this is whether we allow full-screen or not when we check if we have permission to serve ads. This applies to all the changes regarding "ShouldAllow..."

Reply:
Do you have any suggestions? Would `AllowPermission*` work?

Reply:
How about `Has...Permission`, i.e., `HasCatalogPermission`, `HasAllowFullScreenModePermission` etc.?

Reply:
Done.